lib/hierarchical.h

```diff
@@ -334,7 +363,125 @@ class HierarchicalRadiosity {
             for (auto& p : nodes_) {
                 push_pull_radiosity(p, Color());
             }
+
+            // Print progress bar
+            float progress = static_cast<float>(max_iterations_ - iteration) / max_iterations_;
+            const int bar_width = progress * 20;
+            std::cerr << "\rSolving System     "
+                      << "[" << std::string(bar_width, '-')
+                      << std::string(20 - bar_width, ' ') << "] "
+                      << std::setfill(' ') << std::setw(6) << std::fixed
+                      << std::setprecision(2) << (progress * 100.0) << '%';
+            std::cerr.flush();
+        }
+        std::cerr << std::endl;
+    }
+
+    /**
+     * Refine all links in all nodes.
+     * @return true if at least one link has been refined.
+     */
+    bool refine_links() {
+        bool refined = false;
+        for (size_t n = 0; n < nodes_.size(); ++n) {
+            refined |= refine_links(nodes_[n]);
+
+            // Print progress bar
+            float progress = static_cast<float>(n+1) / nodes_.size();
+            const int bar_width = progress * 20;
+            std::cerr << "\rRefining Links     "
+                      << "[" << std::string(bar_width, '-')
+                      << std::string(20 - bar_width, ' ') << "] "
+                      << std::setfill(' ') << std::setw(6) << std::fixed
+                      << std::setprecision(2) << (progress * 100.0) << '%';
+            std::cerr.flush();
+        }
+        std::cerr << std::endl;
+
+        return refined;
+    }
+
+    /**
+     * Refine all links in node p.
+     * @return true if at least one link has been refined.
+     */
+    bool refine_links(Quadnode& p) {
+        bool refined = false;
+
+        // Process all child nodes first.
+        if (!p.is_leaf()) {
+            for (auto& child: p.children) {
+                refined |= refine_links(*child.get());
+            }
+        }
+
+        // Post-order: Process links.
+        // Remove link if we refined. Note: new links might be added that's why
+        // we keep track of the old size.
+        std::vector<Linknode>& links = p.gathering_from;
+        size_t size = links.size();
+        size_t i = 0;
+        while (i < size) {
+            if (refine_link(p, links[i])) {
+                links.erase(links.begin() + i);
```

Comment:
Ich würde hier `std::swap(link.begin() + i, links.begin() + size - 1)` machen. In dem Fall sollte man dann `i` nicht erhöhen. Und dann am Ende der Funktion: `links.erase(links.begin() + size, links.end())`. So vermeidet man das Kopieren der Elemente, wenn man ein Element aus der Mitte löscht.